checks.d/hdfs_datanode.py

```diff
@@ -0,0 +1,173 @@
+hdfs.datanode.num_blocks_failed_to_uncache   The number of failed blocks to remove from cache
+'''
+
+# stdlib
+from urlparse import urljoin
+
+# 3rd party
+import requests
+from requests.exceptions import Timeout, HTTPError, InvalidURL, ConnectionError
+from simplejson import JSONDecodeError
+
+# Project
+from checks import AgentCheck
+
+# JMX URI
+JMX_URI = 'http://localhost:50075'
+
+# Service check names
+JMX_SERVICE_CHECK = 'hdfs.datanode.jmx.can_connect'
+
+# URL Paths
+JMX_PATH = 'jmx'
+
+# Metric types
+GAUGE = 'gauge'
+
+# HDFS bean name
+HDFS_DATANODE_BEAN_NAME = 'Hadoop:service=DataNode,name=FSDatasetState'
+
+# HDFS metrics
+HDFS_METRICS = {
+    'Remaining' : ('hdfs.datanode.dfs_remaining',  GAUGE),
+    'Capacity' :('hdfs.datanode.dfs_capacity', GAUGE),
+    'DfsUsed' : ('hdfs.datanode.dfs_used', GAUGE),
+    'CacheCapacity' : ('hdfs.datanode.cache_capacity', GAUGE),
+    'CacheUsed' : ('hdfs.datanode.cache_used', GAUGE),
+    'NumFailedVolumes' : ('hdfs.datanode.num_failed_volumes', GAUGE),
+    'LastVolumeFailureDate' : ('hdfs.datanode.last_volume_failure_date', GAUGE),
+    'EstimatedCapacityLostTotal' : ('hdfs.datanode.estimated_capacity_lost_total', GAUGE),
+    'NumBlocksCached' : ('hdfs.datanode.num_blocks_cached', GAUGE),
+    'NumBlocksFailedToCache' : ('hdfs.datanode.num_blocks_failed_to_cache', GAUGE),
+    'NumBlocksFailedToUnCache' : ('hdfs.datanode.num_blocks_failed_to_uncache', GAUGE)
+}
+
+class HDFSDataNode(AgentCheck):
+
+    def check(self, instance):
+        jmx_address = instance.get('hdfs_datanode_jmx_uri', JMX_URI)
+
+        # Get metrics from JMX
+        self._hdfs_datanode_metrics(jmx_address, JMX_PATH)
+
+    def _hdfs_datanode_metrics(self, jmx_uri, jmx_path):
+        '''
+        Get HDFS data node metrics from JMX
+        '''
+        response = self._rest_request_to_json(jmx_uri,
+            jmx_path,
+            JMX_SERVICE_CHECK,
+            qry='Hadoop:service=DataNode,name=FSDatasetState')
+
+        beans = response.get('beans', [])
+
+        for bean in beans:
+            bean_name = bean.get('name')
+
+            if bean_name is not None \
+                    and bean_name == HDFS_DATANODE_BEAN_NAME:
+
+                for metric, (metric_name, metric_type) in HDFS_METRICS.iteritems():
+                    metric_value = bean.get(metric)
+
+                    if metric_value is not None:
+                        self._set_metric(metric_name, metric_type, metric_value)
+
+    def _set_metric(self, metric_name, metric_type, value, tags=None, device_name=None):
+        '''
+        Set a metric
+        '''
+        if metric_type == GAUGE:
+            self.gauge(metric_name, value, tags=tags, device_name=device_name)
+        else:
+            self.log.error('Metric type "%s" unknown' % (metric_type))
+
+    def _rest_request_to_json(self, address, object_path, service_name=JMX_SERVICE_CHECK, *args, **kwargs):
```

Comment:
Let's simplify the function signature a bit since we're only using the `address` and `object_path`. Keeping only these 2 parameters should be enough :)